### PR TITLE
[Lang] [ir] Consecutive indices for all snodes

### PR DESCRIPTION
**Pull request by KLozes**
Related issue = #2177
This is main feature of the draft PR #2248. It makes indices for all snodes consecutive. After this I can make a PR to fully deprecate trailing bits.  

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

**Review by k-ye — COMMENTED** [review]
Sorry , just some update to let you know that I'm still reviewing this PR.  I've spent quite some time reviewing the refine coordinates part. From what I understand, the end result of running the `refine_coordinates` from root to leaf is still the same, it's the intermediary results that are changed. If so, would it be possible to separate this part of the change to a smaller PR?

File: taichi/ir/scratch_pad.h
Comment (k-ye) [inline taichi/ir/scratch_pad.h:35]: nit: `coefficients`?

File: taichi/runtime/llvm/runtime.cpp
Comment (k-ye) [inline taichi/runtime/llvm/runtime.cpp:1092]: Could you add a comment explaining that we don't have to refine coordinates, because root snode's num_bits is always 0?

File: taichi/transforms/make_block_local.cpp
Comment (KLozes) [reply taichi/transforms/make_block_local.cpp:149]: Alright, I can throw an 'if' there to check. I imagine it would get optimized out by the final compiler anyways though, right?

File: taichi/codegen/codegen_llvm.cpp
Comment (KLozes) [reply taichi/codegen/codegen_llvm.cpp:1555]: Should I set the rest of the coordinates to null as well: parent_coordinates, current_coordinates, new_coordinates?

Comment (k-ye) [reply taichi/codegen/codegen_llvm.cpp:1555]: Ah yeah, that will be great. We don't have to do this for `new_coordinates` though, because it is a local variable

Comment (k-ye) [reply taichi/codegen/codegen_llvm.cpp:1555]: Ah yeah, that will be great. We don't have to do this for `new_coordinates` though, because it is a local variable

File: tests/python/test_sparse_deactivate.py
Comment (k-ye) [inline tests/python/test_sparse_deactivate.py:28]: Hmm, I don't think this is relevant with this PR. That is, even before this change, changing this to `1` still passes the test.

IIUC, changing to consecutive indices requires us to change how `GlobalPtr` gets lowered to a chain of `GetCh` and `SNodeLookup` stmts. By having `print_ir=True`, you can see that there is a `bit_extract($2) bit_range=[4, 8)` stmt, where `$2` is the indices to be deactivated.  The `bit_range[4,8)` is what is actually causing the non-consecutive indices. @xumingkuan to confirm.

Comment (KLozes) [reply tests/python/test_sparse_deactivate.py:28]: So before this PR `ti.deactivate(ptr, 4)` would deactivate the first block. In fact, any index 0-15 would deactivate the first block. So to maintain the same exact behavior with with consecutive indices, I should have changed it to `ti.deactivate(ptr, 0)`. I just got lucky that `ti.deactivate(ptr, 1)` worked actually.  With this PR `ti.deactivate(ptr, 4)` defiently will not pass. Since it would deactivate the 5th block, which was never activated in the first place!

Comment (KLozes) [reply tests/python/test_sparse_deactivate.py:28]: will change to `ti.deactivate(ptr, 0)`

File: taichi/struct/struct.cpp
Comment (KLozes) [reply taichi/struct/struct.cpp:59]: We shouldn't have to check that anymore! The shape of all snodes are now totally independent of their siblings. See the complex struct test:
``` python
   blk = ti.root.pointer(ti.ij, 4)
   blk.place(w)
   blk.pointer(ti.ij, 2).dense(ti.ij, 2).place(x)
   blk.dense(ti.i, 4).place(y)
   blk.dense(ti.k, 4).place(z)
```
Structures like this are now totally valid

File: tests/python/test_sparse_deactivate.py
Comment (xumingkuan) [reply tests/python/test_sparse_deactivate.py]: Right, `bit_range=[4, 8)` will become `bit_range=[0, 4)` after this PR. `ti.deactivate(ptr, 0)` SGTM!

File: taichi/transforms/make_block_local.cpp
Comment (KLozes) [reply taichi/transforms/make_block_local.cpp:222]: Thanks for the tip!

**Review by k-ye — COMMENTED** [review]
Thank you for moving this forward! I only have  a few nits and test suggestions now

File: tests/python/bls_test_template.py
Comment (KLozes) [reply tests/python/bls_test_template.py:189]: How do I run a cpp test btw?

Comment (k-ye) [reply tests/python/bls_test_template.py:189]: After `make`, there will be a `taichi_cpp_tests` executable. You can just run that and see the results

Comment (KLozes) [reply tests/python/bls_test_template.py:189]: Im sorry, where is the `taichi_cpp_tests` located? I don't see a Makefile over in tests, so I'm guessing you mean after making the project.

File: taichi/transforms/lower_access.cpp
Comment (KLozes) [reply taichi/transforms/lower_access.cpp:85]: Sounds good to me!

File: tests/python/bls_test_template.py
Comment (KLozes) [reply tests/python/bls_test_template.py:189]: Right.. I do not see taichi_cpp_tests exec anywhere! This all I have after make
```
klozes@system76-pc:~/Documents/software/taichi/build$ ls -a
.  ..  CMakeCache.txt  CMakeFiles  cmake_install.cmake  external  libtaichi_core.so  Makefile
```

Comment (k-ye) [reply tests/python/bls_test_template.py:189]: My bad! You need to invoke `cmake` with `-DTI_BUILD_TESTS=ON`...

Comment (KLozes) [reply tests/python/bls_test_template.py:189]: thanks!

**Comment by KLozes** [comment]
@k-ye and @xumingkuan,

Thank you for the valuable comments and reviews! I think I've implemented all the suggestions. I think were just about home free.

**Review by xumingkuan — APPROVED** [review]
Great work, LGTM!

**Comment by xumingkuan** [comment]
Well... I wonder if `ScratchPad::pad_size_linear()` should be multiplied with the coefficients?

**Comment by k-ye** [comment]
> Thank you for the valuable comments and reviews! I think I've implemented all the suggestions. I think were just about home free.

Sorry, I need to go over the entire process of `make_block_local()` again. I believe there are a few issues because of the coefficients.

**Review by k-ye — APPROVED** [review]
> Well... I wonder if `ScratchPad::pad_size_linear()` should be multiplied with the coefficients?

Giving it more thoughts, I believe this is not necessary, and what is done by this PR looks correct. The tricky part is that `ScratchPad` corresponds to the SNode that the BLS wants to cache, not the one being iterated over in the struct-for. Hence, the `pad_size` will be correctly scaled by the coefficients already. As a comparison, `BlockCornerStmt` corresponds to the SNode being iterated over, so we have to manually scale this one by the coefficients.

@KLozes it was such a great pleasure working with you on this PR! This fundamentally touches and changes so many critical parts of Taichi. Learnt a great deal from the work. And thanks @xumingkuan @yuanming-hu  for many rounds of offline discussion!

One last thing: I'm working on a local PR to make `make_block_local()` testable. Could you hold off on that a bit?

File: taichi/struct/struct_llvm.cpp
Comment (KLozes) [reply taichi/struct/struct_llvm.cpp:265]: ****

Comment (KLozes) [reply taichi/struct/struct_llvm.cpp:265]: Yes, trailing bits should always be zero now. I'm planning to totally remove trailing_bits in a subsequent PR.

**Comment by KLozes** [comment]
> @KLozes` it was such a great pleasure working with you on this PR! This fundamentally touches and changes so many critical parts of Taichi. Learnt a great deal from the work. And thanks @xumingkuan @yuanming-hu for many rounds of offline discussion!

@k-ye It was pleasure working with you too! I really enjoy following the great work you guys do on this project. I've learned a ton working on this PR too. It ended up being more complicated than I expected. 

> One last thing: I'm working on a local PR to make make_block_local() testable. Could you hold off on that a bit?

Sure thing, I'll try to keep this PR update to date with the trunk in the mean time.

**Comment by KLozes** [comment]
uhoh @xumingkuan  @k-ye, 

This new optimization pass #2332 broke block local storage. DiffRange cannot handle the left shift operator. I think we realistically only need to support two cases: 
index << const
const << const

So something like this could work
``` cpp
DiffRange operator<<(const DiffRange &a, const DiffRange &b) {
  return DiffRange(
      a.related_() && b.related_() && b.coeff == 0 && b.high - b.low == 1,
      a.coeff << b.low, a.low << b.low, ((a.high - 1) << b.low) + 1);
}
```

**Comment by k-ye** [comment]
> uhoh @xumingkuan @k-ye,
> 
> This new optimization pass #2332 broke block local storage. DiffRange cannot handle the left shift operator. I think we realistically only need to support two cases:
> index << const
> const << const
> 
> So something like this could work
> 
> ```c++
> DiffRange operator<<(const DiffRange &a, const DiffRange &b) {
>   return DiffRange(
>       a.related_() && b.related_() && b.coeff == 0 && b.high - b.low == 1,
>       a.coeff << b.low, a.low << b.low, ((a.high - 1) << b.low) + 1);
> }
> ```

I see. Feel free to add this in a new PR, thanks!

**Comment by xumingkuan** [comment]
> uhoh @xumingkuan @k-ye,
> 
> This new optimization pass #2332 broke block local storage. DiffRange cannot handle the left shift operator. I think we realistically only need to support two cases:
> index << const
> const << const
> 
> So something like this could work
> 
> ```c++
> DiffRange operator<<(const DiffRange &a, const DiffRange &b) {
>   return DiffRange(
>       a.related_() && b.related_() && b.coeff == 0 && b.high - b.low == 1,
>       a.coeff << b.low, a.low << b.low, ((a.high - 1) << b.low) + 1);
> }
> ```

Thanks for pointing this out! `DiffRange operator<<` sounds great to me.

Also, this shows that our tests don't cover this case 😅

**Comment by KLozes** [comment]
I'll make a cpp test for it. :)

File: tests/cpp/transforms/make_block_local_test.cpp
Comment (k-ye) [inline tests/cpp/transforms/make_block_local_test.cpp:135]: IIUC, without the trailing_bits, the `loop_indices_vals` will not be scaled by block_size? (I.e., the looping range is the same as the shape of `struct_for_snode_`)? The block corner itself is not scaled by block_size, either. Instead, `make_bls()` will generate a statement to scale that by coefficients.

Comment (KLozes) [reply tests/cpp/transforms/make_block_local_test.cpp:135]: Yep. exactly right :)

Comment (k-ye) [reply tests/cpp/transforms/make_block_local_test.cpp:135]: Then neither L135-136 nor  `get_block_corner()` needs to multiply by `get_block_size(x)`? I guess we also need to update how `expected_bls_offset_in_bytes` are calculated..

Comment (KLozes) [reply tests/cpp/transforms/make_block_local_test.cpp:135]: Oh sorry I misread your comment. So the coefficients of the scratch pad come directly from the user multiplying the loop index with the block size. This is why in the mpm example, we have to manually rescale the loop index to make it a block_corner before passing it to assume_in_range. 
Also, This test is not a perfect copy of the of mpm example since we are not using the position of the particles to get the central index of the neighborhood to do the scatter/gather. Instead we are just using the loop index, which is just a block index of the grid snode. So we need to rescale this one too, making it also simply a block corner.

Comment (k-ye) [reply tests/cpp/transforms/make_block_local_test.cpp:135]: > So the coefficients of the scratch pad come directly from the user multiplying the loop index with the block size. This is why in the mpm example, we have to manually rescale the loop index to make it a block_corner before passing it to assume_in_range.

Right, I should have better explained the test code: the `loop_indices_vals` array is meant to mock the value of `LoopIndexStmt` for each axis in the `ArithmeticInterpretor`. Similarly, `block_corner_vals` is to mock each `BlockCornerIndexStmt` (see https://github.com/taichi-dev/taichi/blob/22e33a605b5caccc62a02be491a3dc33c2f31c91/tests/cpp/transforms/make_block_local_test.cpp#L170-L177). As you've pointed out, it is the user that is manually multiplying loop index by the block size (L105-108). So `LoopIndexStmt`/`BlockCornerIndexStmt` themselves should not be scaled yet.

We happen to use `loop_indices_vals` and `block_corner_vals` to compute `expected_bls_offset_in_bytes`, which is probably causing lots of confusion. Since the formula of BLS index <--> global indices  has changed, I think we can update how `expected_bls_offset_in_bytes` gets computed. I.e.:

```
expected_bls_offset_in_bytes = sum_i inc_i
inc_i = ( loop_index_i * block_size   # multiplication done by the user
            - block_corner_i * block_size  # multiplication generated by the compiler
            - bls_lower_bounds_i)
```

WDYT?

Comment (KLozes) [reply tests/cpp/transforms/make_block_local_test.cpp]: Ahh yes, I think I see what you mean. the BlockCornerIndexStmt _should_ be returning consecutive indices in this cases since we are looping over a dynamic snode which has zero num_bits for i and j. So it does not get refined to non conecutive indices.

Comment (KLozes) [reply tests/cpp/transforms/make_block_local_test.cpp]: I have implemented your suggestion

File: tests/cpp/analysis/bls_analyzer_test.cpp
Comment (k-ye) [inline tests/cpp/analysis/bls_analyzer_test.cpp:16]: Remove?

File: tests/cpp/analysis/bls_analyzer_test.cpp
Comment (k-ye) [inline tests/cpp/analysis/bls_analyzer_test.cpp:91]: nit: Since we mostly want to test `DiffRange`'s multiplication and SHL, we can create a new test just for `DiffRange`. But I think these are fine to have :-)

**Review by k-ye — APPROVED** [review]
Thank you again for the great works and patience, LGTM!

File: tests/cpp/analysis/bls_analyzer_test.cpp
Comment (KLozes) [reply tests/cpp/analysis/bls_analyzer_test.cpp]: Oops. not sure how that happened. reverted now

File: tests/cpp/analysis/bls_analyzer_test.cpp
Comment (KLozes) [reply tests/cpp/analysis/bls_analyzer_test.cpp:78]: Alright. yeah we don't seem to have any tests specifically for DiffRange yet. I'll figure out how to make some.

Comment (KLozes) [reply tests/cpp/analysis/bls_analyzer_test.cpp:78]: @k-ye tests add!

Comment (k-ye) [reply tests/cpp/analysis/bls_analyzer_test.cpp:78]: Awesome. FYI one of our interns will test this PR with https://github.com/taichi-dev/taichi_elements next week. Let's see how well it works (or what we are missing...)!

**Comment by KLozes** [comment]
The format server is not working on my end. Is anyone else having the same problem?

**Comment by KLozes** [comment]
@k-ye. Would you be able to merge this for me? I cant get the code format and CI to pass for some reason no matter what I do.

**Comment by k-ye** [comment]
> @k-ye. Would you be able to merge this for me? I cant get the code format and CI to pass for some reason no matter what I do.

Sure, I'll be happy to do so. FYI @Jack12xl is helping carry out some final verification.

**Comment by Jack12xl** [comment]
Hi, guys

I test the `rescale_index` with the [Taichi_elements](https://github.com/taichi-dev/taichi_elements/). To clarify, the `taichi`  is compiled with `0.7.20` merged with [this](https://github.com/taichi-dev/taichi/pull/2327) pull request. The [taichi elements](https://github.com/Jack12xl/taichi_elements/tree/final-merge-cons) is at this [commit ](https://github.com/Jack12xl/taichi_elements/tree/final-merge-cons). 

I basically test with the `demo_2d` and `demo_3d`, which is running smoothly after merging the [70](https://github.com/taichi-dev/taichi_elements/pull/70) PR with some fix.

I believe the consecutive index is ready to go !

**Comment by KLozes** [comment]
> I test the rescale_index with the Taichi_elements. To clarify, the taichi is compiled with 0.7.20 merged with this pull request. The taichi elements is at this commit .

Yep that sounds right! Thanks for the review @Jack12xl :).

**Comment by k-ye** [comment]
Sorry I forgot to loop you in. https://github.com/taichi-dev/taichi/pull/2366 has tweaked `BLSAnalyzer` to introduce a `run()` method. Could you update the corresponding tests? Thanks!

File: taichi/analysis/bls_analyzer.cpp
Comment (k-ye) [inline taichi/analysis/bls_analyzer.cpp:54]: s/and/&&, msvc doesn't seem to support `and` 😂

**Review by k-ye — APPROVED** [review]

File: taichi/analysis/bls_analyzer.cpp
Comment (KLozes) [reply taichi/analysis/bls_analyzer.cpp:54]: Oops! I had no idea that was even valid syntax with any compiler! will fix

**Comment by k-ye** [comment]
I think it's fine to temporarily ignore the format issue in this PR. The clang-format on the format server has not played well when there's a change to `export_lang.cpp` for a long time.. We're upgrading the automatic code formatting CI process.

**Comment by KLozes** [comment]
> I think it's fine to temporarily ignore the format issue in this PR

Sounds good to me! Feel free to merge whenever :).

**Comment by k-ye** [comment]
Merged :-) Thank you ALL !! @KLozes @xumingkuan @Jack12xl @yuanming-hu

**Pull request merged**
